File: java/datastructure/array/leetcode/ContainsDuplicate.java
package datastructure.array.leetcode;

import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

/**
 * @author roseduan
 * @time 2020/10/24 9:16 上午
 * @description 存在重复元素
 */
public class ContainsDuplicate {

    /**
     * 排序
     */
    public boolean containsDuplicate1(int[] nums) {
        Arrays.sort(nums);
        for (int i = 0; i < nums.length - 1; i++) {
            if (nums[i] == nums[i + 1]) {
                return true;
            }
        }
        return false;
    }

    /**
     * 使用哈希表
     */
    public boolean containsDuplicate2(int[] nums) {
        Set<Integer> set = new HashSet<>();
        for (int n : nums) {
            if (set.contains(n)) {
                return true;
            }
            set.add(n);
        }
        return false;
    }
}

File: cpp/array/ contains_duplicate.cpp
#include "vector"
#include "iostream"
#include "unordered_set"

using namespace std;

class Solution {
public:

    // 使用一个set
    bool containsDuplicate1(vector<int>& nums) {
        unordered_set<int> s;
        for (int i : nums) {
            if (s.count(i)) {
                return true;
            }
            s.insert(i);
        }
        return false;
    }

    // 排序
    bool containsDuplicate2(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        for (int i = 1; i < nums.size(); ++i) {
            if (nums[i] == nums[i-1]) {
                return true;
            }
        }
        return false;
    }
};
